File: src/Diplom/Node.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package Diplom;

import java.util.ArrayList;

/**
 *
 * @author osboxes
 */
public class Node {

    public ArrayList<PointDim> PointSet = new ArrayList<>();
    public PointDim left = new PointDim();
    public PointDim right = new PointDim();
    public ArrayList<Node> Pairs = new ArrayList<>();

    public Node() {
    }

    public Node(ArrayList<PointDim> PointSet) {
        this.PointSet = PointSet;
    }

    public Node(PointDim left, PointDim right) {
        this.left = left;
        this.right = right;
    }

    public void UpdateBoundingValues() {
        if (!this.IsEmpty()) {
            ResetBoundingValues();
            if (PointSet.size() < 100) {//this is kostill for faster 0.5 second
                for (PointDim point : PointSet) {
                    CheckBoundingValues(point);
                }
            } else {
                int blaStep;
                if (PointSet.size() > 1000) {
                    if (PointSet.size() > 10000) {
                        if (PointSet.size() > 100000) {
                            blaStep = 10000;
                        } else {
                            blaStep = 1000;
                        }
                    } else {
                        blaStep = 100;
                    }
                } else {
                    blaStep = 10;
                }
                for (int bla = 0; bla < PointSet.size(); bla += blaStep) {
                    CheckBoundingValues(PointSet.get(bla));
                }
            }
        }
    }

    private void ResetBoundingValues() {
        for (int i = 0; i < Base.dimension; ++i) {
            this.left.X[i] = 1;
            this.right.X[i] = 0;
        }
    }

    private void CheckBoundingValues(PointDim point) {
        for (int i = 0; i < Base.dimension; ++i) {
            if (this.left.X[i] > point.X[i]) {
                this.left.X[i] = point.X[i];
            }
            if (this.right.X[i] < point.X[i]) {
                this.right.X[i] = point.X[i];
            }
        }
    }

    public boolean OnlyOneElement() {
        return PointSet.size() == 1;
    }

    public boolean IsEmpty() {
        return PointSet.isEmpty();
    }

    public Node SplitSelf() {
//        d0 = System.currentTimeMillis();
        Node NewNode = new Node();
        ArrayList<PointDim> NewPointSet = new ArrayList<>(PointSet.size() >> 1);
        NewNode.PointSet = new ArrayList<>(PointSet.size() >> 1);
        float middle;
        int maxSideIndex = 0;
        for (int i = 1; i < Base.dimension; ++i) {
            if (this.right.X[maxSideIndex] - this.left.X[maxSideIndex]
                    < this.right.X[i] - this.left.X[i]) {
                maxSideIndex = i;
            }
        }
        middle = (right.X[maxSideIndex] + left.X[maxSideIndex]) / 2f;
//        d1 += System.currentTimeMillis() - d0;
        for (PointDim point : PointSet) {
            if (point.X[maxSideIndex] > middle) {
                NewNode.PointSet.add(point);
            } else {
                NewPointSet.add(point);
            }
        }
        PointSet = NewPointSet;
//        d2 += System.currentTimeMillis() - d0;
        this.UpdateBoundingValues();
        NewNode.UpdateBoundingValues();
//        d3 += System.currentTimeMillis() - d0;
        return NewNode;
    }

    public void Clone(Node node) {
        this.PointSet = node.PointSet;
        this.left = node.left;
        this.right = node.right;
        this.Pairs = node.Pairs;
    }
}

File: Node.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Drawing;
using System.Diagnostics;

namespace Diplom
{
    public class Node
    {
        public List<PointDim> PointSet = new List<PointDim>();
        public float[] left = new float[Base.DIMENSION];
        public float[] right = new float[Base.DIMENSION];
        public HashSet<Node> Pairs = new HashSet<Node>();
        public Node()
        {
        }
        public Node(List<PointDim> PointSet)
        {
            this.PointSet = PointSet;
        }
        public Node(float[] left, float[] right)
        {
            this.left = left;
            this.right = right;
        }
        public void ResetBoundingValues()
        {
            if (!this.IsEmpty())
            {
                for (int i = 0; i < Base.DIMENSION; ++i)
                {
                    this.left[i] = 1;
                    this.right[i] = 0;
                }
                foreach (PointDim point in PointSet)
                {
                    for (int i = 0; i < Base.DIMENSION; ++i)
                    {
                        if (this.left[i] > point.X[i])
                        {
                            this.left[i] = point.X[i];
                        }
                        if (this.right[i] < point.X[i])
                        {
                            this.right[i] = point.X[i];
                        }
                    }
                }
            }
        }
        public bool OnlyOneElement()
        {
            return (PointSet.Count == 1);
        }
        public bool IsEmpty()
        {
            return (PointSet.Count == 0);
        }
        public void AddPoint(PointDim point)
        {
            PointSet.Add(point);
        }
        public Node SplitSelf()
        {
            Node NewNode = new Node();
            List<PointDim> NewPointSet = new List<PointDim>();
            float middle;
            int maxSideIndex = 0;
            for (int i = 1; i < Base.DIMENSION; ++i)
            {
                if (this.right[maxSideIndex] - this.left[maxSideIndex] <
                    this.right[i] - this.left[i])
                {
                    maxSideIndex = i;
                }
            }
            middle = (right[maxSideIndex] + left[maxSideIndex]) / 2;
            foreach (PointDim point in PointSet)
            {
                if (point.X[maxSideIndex] > middle)
                {
                    NewNode.AddPoint(point);
                }
                else
                {
                    NewPointSet.Add(point);
                }
            }
            PointSet = NewPointSet;
            this.ResetBoundingValues();
            NewNode.ResetBoundingValues();
            return NewNode;
        }
        public void Clone(Node node)
        {
            this.PointSet = node.PointSet;
            this.left = node.left;
            this.right = node.right;
            this.Pairs = node.Pairs;
        }
        public PointDim GetCenter()
        {
            PointDim Center = new PointDim();
            for (int i = 0; i < Base.DIMENSION; ++i)
            {
                Center.X[i] = (left[i] + right[i]) / 2;
            }
            return Center;
        }
        public PointDim GetLeft()
        {
            PointDim VeryLeft = new PointDim();
            for (int i = 0; i < Base.DIMENSION; ++i)
            {
                VeryLeft.X[i] = left[i];
            }
            return VeryLeft;
        }
    }
}
